rfcs/syntax-list-comprehensions.md

```diff
@@ -0,0 +1,78 @@
+## Summary
+
+Introduce a form of list comprehension using `n for-in-do` syntax.
+
+## Motivation
+
+List comprehensions would bring several benefits and prevent code smell.
+In Lua you are encouraged to not modify tables during traversal.
+
+When traversing a table to exclude all the odd numbers you'd be creating a large statement to get rid of them, with list comprehensions this could be shortened and cleaner.
+
+## Design
+
+To solve these problems, I propose a `n for-in-do` expression form that is syntactically similar to a for statement, but lacks terminating `end`.
+
+The `n for-in-do` expression must match ``<identifier> for <identifier> in <expr> do``.
+A ``if <expr>`` expression might be required.
+
+Example:
+```lua
+-- normal
+local t = {1,2,3,4,5,6,7,8,9}
+local onlyEven = {}
+for i,n in pairs(t) do
+    if n%2 == 0 then 
+      table.insert(onlyEven, n)
+    end
+end
+
+-- list comprehensions
+local t = {1,2,3,4,5,6,7,8,9}
+local onlyEven = {n for n in t do if n%2 == 0} 
+```
```

Comment:
Is this a normal expression or just one that can be used in a table constructor?
```lua
local t = {"a","b","c"}
return -- can a list comprehension be used here to return all values within `t` that are not "b"?
```
What does the first identifier before the `for` represent? Is it a just the name specified in-between `for` and `in` stated again, or does it refer to something else?

The syntax of the for statement is `for <namelist> in <explist> do <block> end`, so it isn't very syntactically similar to the for statement. The example provided uses only a table, so it doesn't work like a for statement. The example doesn't make it very clear what is being referred to by `n`, is it the key or the value?

It would be helpful if the semantics and syntax were better explained, and if better examples were provided to show why a feature like this is desired.

Reply:
It could be used as a normal expression, I just listed it as list comprehensions which would limit it to arrays